lib/LaTeXML/Post.pm

```diff
@@ -621,6 +623,28 @@ sub addCrossrefs {
   } }
   return; }
 
+sub mathIsParsed {
+  my ($doc, $math) = @_;
+  # when MathParser fails, it marks the failed fragment with class="unparsed"
+  my $unparsed = $math && $doc->findnode('descendant::*[@class="unparsed"]', $math);
+  my @unparsed_in_duals = $unparsed ? $doc->findnodes('descendant::*[self::ltx:XMDual]/descendant::*[@class="unparsed"]', $math) : ();
+  # simple case - a regular expression failed to parse,
+  return 0 if $unparsed && !@unparsed_in_duals;
+# complex case - an XMDual failed to parse
+# We need to be careful about allowed parse failures, namely:
+# The presentation branch of an XMDual (second child) are not used in Content post-processors, even if ungrammatical.
+# However, if an "unparsed" class is inside the content child, or inside an XMRef branch, it is indeed a failure to parse.
+  for my $node (@unparsed_in_duals) {
+    my $content_branch = $doc->findnode('ancestor::ltx:XMDual/*[1]');
+    return 0 unless mathIsParsed($doc, $content_branch);
+    my @ref_nodes = $doc->findnodes('descendant::*[self::ltx:XMRef]', $content_branch);
+    for my $ref_node (@ref_nodes) {
+      my $rnode = $doc->realizeXMNode($ref_node);
+      # recursively check if the ref node is parsed, using the same rules, and fail if not
+      return 0 unless mathIsParsed($doc, $rnode); } }
+  # all looks nominal
+  return 1; }
+
```

Comment:
Does that actually work? :>  I had in mind a more pedestrian:
```
sub mathIsParsed {
  my ($doc, $node) = @_;
  my $tag = $doc->getNodeQName($node);
  if (($tag eq 'ltx:XMWrap') || ($tag eq 'ltx:XMArg')) {
    return 0; }
  elsif (($tag eq 'ltx:XMTok') || ($tag eq 'ltx:XMText') || ($tag eq 'ltx:XMHint')) {
    return 1; }
  elsif ($tag eq 'ltx:XMRef') {
    return mathIsParsed($doc, $doc->realizeXMNode($node)); }
  elsif ($tag eq 'ltx:XMDual') {
    my ($content, $presentation) = element_nodes($node);
    return mathIsParsed($doc, $content); }
  else {
    foreach my $child (element_nodes($node)) {
      return 0 unless mathIsParsed($doc, $child); }
    return 1; } }
```
Untested, but should be close.

Reply:
Wouldn't this perl recursion get pretty slow on large formulas in even larger documents? arXiv even has examples of documents that time out in some of those cases.

My initial attempt was to get a handle on all possible paths for a parse failure already in MathParser and just mark the top-level math element with a class, so that the check in post-processing is instant. Although I didn't manage that, since the dual cases where hard to get a handle on...